### PR TITLE
Pv2 625 levy funding source to process refunds first

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using SFA.DAS.Payments.FundingSource.Domain.Interface;
+using SFA.DAS.Payments.RequiredPayments.Messages.Events;
+
+namespace SFA.DAS.Payments.FundingSource.Domain.Services
+{
+    public class SortableKeyGenerator : IGenerateSortableKeys
+    {
+        public string Generate(ApprenticeshipContractType1RequiredPaymentEvent requiredPayment)
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs]: There is no need to pass in an object here, just pass in Amount, Uln, and Priority. Then we can just have various test cases of input to the expected result.

File: src/SFA.DAS.Payments.FundingSource.Domain.UnitTests/GivenASortableKeyGenerator/WhenCallingGenerate.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            keys.Add(refundKey);
+
+            keys.Sort();
+
+            keys[0].Should().Be(refundKey);
+            keys[1].Should().Be(paymentKey);
+        }
+
+        [TestFixture]
+        public class AndAllEventsArePayments : WhenCallingGenerate
```
Comment (samuel-dev-dba) [inline src/SFA.DAS.Payments.FundingSource.Domain.UnitTests/GivenASortableKeyGenerator/WhenCallingGenerate.cs]: Where is the test to handle is not refund but same Priority but different ULN

File: src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs (hunk 1 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs]: We had a very long discussion about not wanting to pass too many parameters. And *all* of these parameters come from the object...

Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs]: I don't think that applies to pass in 3 parameters to a method.

File: src/SFA.DAS.Payments.FundingSource.Domain.UnitTests/GivenASortableKeyGenerator/WhenCallingGenerate.cs (hunk 2 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.FundingSource.Domain.UnitTests/GivenASortableKeyGenerator/WhenCallingGenerate.cs]: ok

File: src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs (hunk 1 above)
Comment (carlbm) [reply src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs]: Changed

File: src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using SFA.DAS.Payments.FundingSource.Domain.Interface;
+
+namespace SFA.DAS.Payments.FundingSource.Domain.Services
+{
+    public class SortableKeyGenerator : IGenerateSortableKeys
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.Domain/Services/SortableKeyGenerator.cs]: "IGenerateSortableKeys" is funny but inconsistent with the class name. Can it be a boring "ISortableKeyGenerator" or something?

File: src/SFA.DAS.Payments.FundingSource.Application.UnitTests/Service/ContractType1RequiredPaymentEventFundingSourceServiceTest.cs [hunk 4]
```diff
@@ -74,14 +78,17 @@ public async Task TestRegisterFirstRequiredPayment()
...
+                Learner = new Learner(),
             };
 
-            var key = string.Concat("000001-000000-", requiredPaymentEvent.EventId.ToString());
+            var key = GenerateSortableKey(requiredPaymentEvent);
 
             eventCacheMock.Setup(c => c.Add(key, requiredPaymentEvent, CancellationToken.None)).Returns(Task.CompletedTask).Verifiable();
             keyCacheMock.Setup(c => c.TryGet("keys", CancellationToken.None)).ReturnsAsync(() => new ConditionalValue<List<string>>(false, null)).Verifiable();
             keyCacheMock.Setup(c => c.AddOrReplace("keys", It.Is<List<string>>(list => list.Count == 1 && list[0] == key), CancellationToken.None)).Returns(Task.CompletedTask).Verifiable();
+            sortableKeysMock.Setup(x => x.Generate(It.IsAny<decimal>(), It.IsAny<int>(), It.IsAny<DateTime>(), It.IsAny<long>(), It.IsAny<Guid>()))
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.Application.UnitTests/Service/ContractType1RequiredPaymentEventFundingSourceServiceTest.cs]: This is not testing much, needs to check on parameters and be verifiable.

File: src/SFA.DAS.Payments.FundingSource.Application.UnitTests/Service/ContractType1RequiredPaymentEventFundingSourceServiceTest.cs [hunk 5]
```diff
@@ -97,14 +104,17 @@ public async Task TestRegisterSubsequentRequiredPayment()
...
+                Learner = new Learner(),
             };
 
-            var key = string.Concat("000004-000002-", requiredPaymentEvent.EventId.ToString());
+            var key = GenerateSortableKey(requiredPaymentEvent);
 
             eventCacheMock.Setup(c => c.Add(key, requiredPaymentEvent, CancellationToken.None)).Returns(Task.CompletedTask).Verifiable();
             keyCacheMock.Setup(c => c.TryGet("keys", CancellationToken.None)).ReturnsAsync(() => new ConditionalValue<List<string>>(true, keys)).Verifiable();
             keyCacheMock.Setup(c => c.AddOrReplace("keys", It.Is<List<string>>(list => list.Count == 3 && list[2] == key), CancellationToken.None)).Returns(Task.CompletedTask).Verifiable();
+            sortableKeysMock.Setup(x => x.Generate(It.IsAny<decimal>(), It.IsAny<int>(), It.IsAny<DateTime>(), It.IsAny<long>(), It.IsAny<Guid>()))
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.Application.UnitTests/Service/ContractType1RequiredPaymentEventFundingSourceServiceTest.cs]: This is not testing much, needs to check on parameters and be verifiable.

File: src/SFA.DAS.Payments.FundingSource.Application/Services/ContractType1RequiredPaymentEventFundingSourceService.cs [hunk 6]
```diff
@@ -25,27 +25,30 @@ public class ContractType1RequiredPaymentEventFundingSourceService : IContractTy
...
             this.levyAccountRepository = levyAccountRepository ?? throw new ArgumentNullException(nameof(levyAccountRepository));
             this.levyBalanceService = levyBalanceService;
+            this.sortableKeys = sortableKeys;
         }
 
         public async Task AddRequiredPayment(ApprenticeshipContractType1RequiredPaymentEvent paymentEvent)
         {
             var keys = await GetKeys().ConfigureAwait(false);
-            var key = GenerateSortableKey(paymentEvent.EventId, paymentEvent.Priority, keys.Count);
+            var key = sortableKeys.Generate(paymentEvent.AmountDue, paymentEvent.Priority, DateTime.MaxValue, paymentEvent.Learner.Uln, paymentEvent.EventId);
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.FundingSource.Application/Services/ContractType1RequiredPaymentEventFundingSourceService.cs]: Why are you not adding and using AgreementDate here?

**Pull request merged**